docs/source/model_library.rst

```diff
@@ -0,0 +1,550 @@
+
+It is generally recommended to expose ``on_disk`` in the ``Step.spec``
+allowing the `~stpipe.step.Step` to generate an :ref:`library_on_disk`
+`~stpipe.library.AbstractModelLibrary`:
+
+.. code-block:: pycon
+
+   >>> class MyStep(Step):
+   ...     spec = """
+   ...         on_disk = boolean(default=False)  # keep models "on disk" to reduce RAM usage
+   ...     """
+
+.. NOTE::
+   As mentioned in :ref:`library_on_disk` at no point will the input files
+   referenced in the association be modified. However, the above pattern
+   does allow ``Step.process`` to "modify" ``init`` when
+   ``init`` is a `~stpipe.library.AbstractModelLibrary` (the models
+   in the library will not be copied).
+
+``Step.process`` can extend the above pattern to
+support additional inputs (for example a single
+`~stpipe.datamodel.AbstractDataModel` or filename containing
+a `~stpipe.datamodel.AbstractDataModel`) to allow more
+flexible data processings. Although some consideration
+should be given to how to handle input that does not
+contain association metadata. Does it make sense
+to construct a `~stpipe.library.AbstractModelLibrary` when the
+association metadata is made up? Alternatively, is
+it safer (less prone to misattribution of metadata)
+to have the step process the inputs separately
+(more on this below)?
+
+.. _library_isolated_processing:
+
+Isolated Processing
+-------------------
+
+Let's say we have a `~stpipe.step.Step`, ``flux_calibration``
+that performs an operation that is only concerned with the data
+for a single `~stpipe.datamodel.AbstractDataModel` at a time.
+This step applies a function ``calibrate_model_flux`` that
+accepts a single `~stpipe.datamodel.AbstractDataModel` and index as an input.
+It's ``Step.process`` function can make good use of
+`~stpipe.library.AbstractModelLibrary.map_function` to apply
+this method to each model in the library.
+
+.. code-block:: pycon
+
+   >>> class FluxCalibration(Step):
+   ...     spec = "..." # use spec defined above
+   ...     def process(self, init):
+   ...         # see input pattern described above
+   ...         # list is used here to consume the generator produced by map_function
+   ...         list(library.map_function(calibrate_model_flux))
+   ...         return library
+
+.. _library_troubleshooting:
+
+Troubleshooting
+===============
+
+.. _library_closed_library_error:
+
+ClosedLibraryError
+------------------
+
+.. code-block:: pycon
+
+   >>> model = library.borrow(0)
+
+   ClosedLibraryError: ModelLibrary is not open
+
+The library must be "open" (used in a ``with`` statement) before
+a model can be borrowed. This is important for keeping track of
+which models were possibly modified.
+
+This error can be avoided by "opening" the library before calling
+`~stpipe.library.AbstractModelLibrary.borrow` (and being sure to call
+`~stpipe.library.AbstractModelLibrary.shelve`, more on that below):
```

Comment:
```suggestion
`~stpipe.library.AbstractModelLibrary.shelve`):
```
my logic is that it's unlikely someone will be reading this section linearly - it's likely to be a reference if someone runs into these errors.

Random note, but I'm impressed by how helpful the various error messages have been during my own development with ModelLibrary - nice work!